File: Earth_Library/scripts/store_to_library.mjs
import { mkdirSync, readFileSync, appendFileSync, writeFileSync, statSync } from "node:fs";
import { join } from "node:path";

const root = process.env.CURSOR_PROJECT_DIR || process.cwd();
const libRoot = join(root, "Earth_Library");
const cardsPath = join(libRoot, "cards.jsonl");
const indexPath = join(libRoot, "library_index.json");
const relPath = join(libRoot, "relations.jsonl");
const queuePath = join(libRoot, "Review_Queue.md");
const cfgPath = join(libRoot, "System", "ingest_config.json");
const tagPath = join(libRoot, "System", "tag_dictionary.json");

const args = Object.fromEntries(
  process.argv.slice(2).map((v, i, a) => (v.startsWith("--") ? [v.slice(2), a[i + 1]] : null)).filter(Boolean)
);
const title = args.title || "untitled";
let content = "";
if (args["content-file"]) {
  content = readFileSync(args["content-file"], "utf8");
} else if (args.content != null && args.content !== "") {
  content = String(args.content);
}
if (!content) {
  process.stdout.write(JSON.stringify({ ok: false, error: "必须提供 --content 或 --content-file" }));
  process.exit(1);
}
const type = args.type || "知识记录";
const source = args.source || "对话沉淀";
const sourceMode = args.source_mode || "conversation";
const sourceUrl = args.source_url || "";
const sourcePath = args.source_path || "";
const confidence = args.confidence || "中";
const keywords = args.keywords || "";
const notionPageId = args["notion-page-id"] || "";

const cfg = JSON.parse(readFileSync(cfgPath, "utf8"));
const tagCfg = JSON.parse(readFileSync(tagPath, "utf8"));
const maxTags = Number(tagCfg.recommended_tag_count?.default || 5);

function slugify(text) {
  return text.trim().replace(/[^\w\u4e00-\u9fff-]+/g, "-").replace(/-{2,}/g, "-").replace(/^-|-$/g, "") || "untitled";
}
function tokenize(text) {
  return new Set(text.split(/[,，\s]+/).map((x) => x.trim().toLowerCase()).filter(Boolean));
}
function parseTags(raw) {
  if (typeof raw === "string") return raw.replace(/，/g, ",").split(",").map((s) => s.trim()).filter(Boolean);
  if (Array.isArray(raw)) return raw.flatMap((item) => parseTags(item));
  return [];
}
function inferTags(payload, maxTags) {
  const text = payload.toLowerCase();
  const tags = [];
  for (const item of tagCfg.tags || []) {
    const name = item.name;
    const triggers = item.triggers || [];
    if (!name) continue;
    if (triggers.some((t) => text.includes(String(t).toLowerCase()))) tags.push(name);
  }
  return [...new Set(tags)].slice(0, maxTags);
}

const now = new Date();
const date = now.toISOString().slice(0, 10);
const ts = now.toISOString().replace(/[-:TZ.]/g, "").slice(0, 14);
const newId = `${ts}_${slugify(title)}`;
const tags = inferTags([title, content, type, source, sourceUrl, sourceMode, confidence, keywords].join(" "), maxTags);

mkdirSync(libRoot, { recursive: true });

// Build card object
const card = {
  id: newId,
  title,
  type,
  confidence,
  tags: tags.join(","),
  keywords,
  source,
  source_url: sourceUrl,
  source_mode: sourceMode,
  source_path: sourcePath,
  notion_page_id: notionPageId || "",
  lifecycle: "阶段",
  created: date,
  body_md: content,
};

// Append to cards.jsonl
appendFileSync(cardsPath, JSON.stringify(card) + "\n", "utf8");

// Update library_index.json
let index = { cards: [] };
try { index = JSON.parse(readFileSync(indexPath, "utf8")); } catch {}
index.cards.push({
  id: newId,
  title,
  type,
  source,
  date,
  keywords: keywords ? keywords.split(",").map((k) => k.trim()).filter(Boolean) : [],
  confidence,
});
writeFileSync(indexPath, JSON.stringify(index, null, 2), "utf8");

// Compute relations
const newKw = tokenize(content);
newKw.add(title.toLowerCase());
const newTagSet = new Set(tags.map((t) => t.toLowerCase()));

// Load existing cards for relation computation
let allCards = [];
if (statSync(cardsPath, { throwIfNoEntry: false })?.isFile()) {
  allCards = readFileSync(cardsPath, "utf8").split("\n").filter((l) => l.trim()).map((l) => {
    try { return JSON.parse(l); } catch { return null; }
  }).filter(Boolean);
}

let relatedCount = 0;
for (const existing of allCards) {
  const eid = existing.id || "";
  if (!eid || eid === newId) continue;

  const ekw = tokenize((existing.body_md || "") + " " + (existing.title || ""));
  const eTags = new Set(parseTags(existing.tags || []).map((t) => t.toLowerCase()));

  // 关键词相交
  if ([...newKw].some((k) => ekw.has(k))) {
    appendFileSync(relPath,
      JSON.stringify({ d: date, s: newId, t: eid, r: "关键词相交", x: "自动关联（共享关键词）" }) + "\n",
      "utf8");
    relatedCount++;
  }

  // 标签相交
  if ([...newTagSet].some((t) => eTags?.has(t))) {
    appendFileSync(relPath,
      JSON.stringify({ d: date, s: newId, t: eid, r: "标签相交", x: "自动关联（共享标签）" }) + "\n",
      "utf8");
    relatedCount++;
  }

  // 冲突检测
  if (
    (existing.body_md || "").includes("# Summary") &&
    (existing.body_md || "").includes(title) &&
    (content.includes("冲突") || content.includes("相反"))
  ) {
    appendFileSync(relPath,
      JSON.stringify({ d: date, s: newId, t: eid, r: "冲突", x: "新增内容包含冲突语义，请人工复核" }) + "\n",
      "utf8");
    appendFileSync(queuePath,
      `\n| ${date} | \`${newId}\` | 冲突待复核 | 与 \`${eid}\` 存在冲突标记语义 | 人工确认口径并修订 | 待处理 |`,
      "utf8");
  }
}

// Low confidence check
if ((cfg.quality_rules?.low_confidence_values || []).includes(confidence)) {
  appendFileSync(queuePath,
    `\n| ${date} | \`${newId}\` | 低置信度 | 该条目标记为低置信度 | 补充来源与证据后复核 | 待处理 |`,
    "utf8");
}

process.stdout.write(JSON.stringify({
  ok: true,
  action: "created",
  id: newId,
  related_count: relatedCount,
  tag_max: maxTags,
}));

File: Earth_Library/scripts/store_to_library.py
from __future__ import annotations

import argparse
import json
import os
import re
from datetime import datetime
from pathlib import Path

from el_parsers import (
    format_notion_page_id_display,
    load_jsonl,
    normalize_notion_page_id,
    parse_tags,
)
from notion_ingest_dedupe import (
    find_card_by_notion_id_jsonl,
    local_supplement_is_non_empty,
    merge_details_inner,
    replace_details_inner,
    wrap_notion_export,
)

_SCRIPT_DIR = Path(__file__).resolve().parent
ROOT = Path(os.environ.get("CURSOR_PROJECT_DIR", _SCRIPT_DIR.parents[2]))
LIB_ROOT = ROOT / "Earth_Library"
CARDS_JSONL = LIB_ROOT / "cards.jsonl"
INDEX_JSON = LIB_ROOT / "library_index.json"
RELATIONS_JSONL = LIB_ROOT / "relations.jsonl"
QUEUE = LIB_ROOT / "Review_Queue.md"
CFG = LIB_ROOT / "System" / "ingest_config.json"
TAG_DICT = LIB_ROOT / "System" / "tag_dictionary.json"


def slugify(text: str) -> str:
    s = re.sub(r"[^\w\u4e00-\u9fff-]+", "-", text.strip())
    return re.sub(r"-{2,}", "-", s).strip("-") or "untitled"


def tokenize(text: str) -> set[str]:
    return {x.strip().lower() for x in re.split(r"[,，\s]+", text) if x.strip()}


def infer_tags(payload: str, max_tags: int) -> list[str]:
    data = json.loads(TAG_DICT.read_text(encoding="utf-8"))
    tags: list[str] = []
    text = payload.lower()
    for item in data.get("tags", []):
        name = item.get("name")
        triggers = item.get("triggers", [])
        if not name:
            continue
        if any(t.lower() in text for t in triggers):
            tags.append(name)
    out: list[str] = []
    for t in tags:
        if t not in out:
            out.append(t)
    return out[:max_tags]


def load_index() -> list[dict]:
    """读取 library_index.json，返回 cards 数组。不存在则返回空列表。"""
    if not INDEX_JSON.exists():
        return []
    data = json.loads(INDEX_JSON.read_text(encoding="utf-8"))
    return data.get("cards", [])


def save_index(cards: list[dict]) -> None:
    """覆盖写入 library_index.json。"""
    INDEX_JSON.parent.mkdir(parents=True, exist_ok=True)
    INDEX_JSON.write_text(json.dumps({"cards": cards}, ensure_ascii=False, indent=2), encoding="utf-8")


def compute_relations(
    *,
    new_id: str,
    new_body: str,
    new_title: str,
    new_tags: list[str],
    existing_cards: list[dict],
    date: str,
) -> tuple[list[dict], list[str]]:
    """计算新卡片与已有卡片的关联。返回 (relation_rows, queue_lines)。"""
    new_kw = tokenize(new_body)
    new_kw.add(new_title.lower())
    new_tag_set = set(t.lower() for t in new_tags)

    relation_rows: list[dict] = []
    queue_lines: list[str] = []

    for card in existing_cards:
        card_id = card.get("id", "")
        if not card_id:
            continue
        card_body = card.get("body_md", "")
        card_title = card.get("title", "")
        card_kw = tokenize(card_body)
        card_kw.add(card_title.lower())
        card_tag_set = set(t.lower() for t in parse_tags(card.get("tags", [])))

        # 关键词相交
        if new_kw and card_kw and (new_kw & card_kw):
            relation_rows.append({
                "d": date,
                "s": new_id,
                "t": card_id,
                "r": "关键词相交",
                "x": "自动关联（共享关键词）",
            })

        # 标签相交
        if new_tag_set and card_tag_set and (new_tag_set & card_tag_set):
            relation_rows.append({
                "d": date,
                "s": new_id,
                "t": card_id,
                "r": "标签相交",
                "x": "自动关联（共享标签）",
            })

        # 冲突检测
        if ("# Summary" in card_body and new_title in card_body) and ("冲突" in new_body or "相反" in new_body):
            relation_rows.append({
                "d": date,
                "s": new_id,
                "t": card_id,
                "r": "冲突",
                "x": "新增内容包含冲突语义，请人工复核",
            })
            queue_lines.append(
                f"\n| {date} | `{new_id}` | 冲突待复核 | 与 `{card_id}` 存在冲突标记语义 | 人工确认口径并修订 | 待处理 |"
            )

    return relation_rows, queue_lines


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--title", required=True)
    parser.add_argument("--content", default=None, help="正文；与 --content-file 二选一")
    parser.add_argument(
        "--content-file",
        dest="content_file",
        default=None,
        help="从 UTF-8 文件读取正文（优先于 --content，用于避免命令行过长）",
    )
    parser.add_argument("--type", default="知识记录")
    parser.add_argument("--source", default="对话沉淀")
    parser.add_argument("--source_mode", default="conversation")
    parser.add_argument("--source_url", default="")
    parser.add_argument("--source_path", default="")
    parser.add_argument("--confidence", default="中")
    parser.add_argument("--keywords", default="")
    parser.add_argument(
        "--notion-page-id",
        dest="notion_page_id",
        default=None,
        help="Notion 页面 ID 或含 ID 的 URL",
    )
    args = parser.parse_args()

    if args.content_file:
        content = Path(args.content_file).read_text(encoding="utf-8")
    elif args.content is not None:
        content = args.content
    else:
        parser.error("必须提供 --content 或 --content-file")

    cfg = json.loads(CFG.read_text(encoding="utf-8"))
    tag_cfg = json.loads(TAG_DICT.read_text(encoding="utf-8"))
    max_tags = int(tag_cfg.get("recommended_tag_count", {}).get("default", 5))

    now = datetime.now()
    date = now.strftime("%Y-%m-%d")
    ts = now.strftime("%Y%m%d-%H%M%S")
    slug = slugify(args.title)
    new_id = f"{ts}_{slug}"

    inferred_tags = infer_tags(
        " ".join([
            args.title, content, args.type, args.source,
            args.source_url, args.source_mode, args.confidence, args.keywords,
        ]),
        max_tags,
    )

    nid_norm: str | None = None
    if args.notion_page_id:
        nid_norm = normalize_notion_page_id(args.notion_page_id)
        if not nid_norm:
            parser.error(f"无法解析 Notion 页面 ID: {args.notion_page_id!r}")

    use_notion_dedupe = bool(nid_norm and args.source_mode == "notion_page")
    notion_display = format_notion_page_id_display(nid_norm) if nid_norm else ""

    action = "created"

    # 处理已有卡片更新（Notion 去重）
    existing = None
    if use_notion_dedupe:
        existing = find_card_by_notion_id_jsonl(CARDS_JSONL, nid_norm)

    cards_list = load_jsonl(CARDS_JSONL)
    index_cards = load_index()

    if existing is not None:
        old_body = existing.get("body_md", "")
        fm = existing.copy()
        old_details = old_body or ""

        if local_supplement_is_non_empty(old_details):
            new_details = merge_details_inner(old_details, content)
            action = "merged"
        else:
            new_details = replace_details_inner(old_details, content)
            action = "replaced"

        # 更新正文
        existing["body_md"] = new_details
        existing["source"] = args.source
        existing["source_mode"] = args.source_mode
        existing["source_url"] = args.source_url
        existing["confidence"] = args.confidence
        existing["tags"] = ",".join(inferred_tags)
        existing["keywords"] = args.keywords

        # 计算关联
        relation_rows, queue_lines = compute_relations(
            new_id=existing["id"],
            new_body=new_details,
            new_title=existing.get("title", ""),
            new_tags=inferred_tags,
            existing_cards=[c for c in cards_list if c.get("id") != existing.get("id")],
            date=date,
        )

        # 写回 cards.jsonl
        _write_jsonl(CARDS_JSONL, cards_list)

        # 追加关联
        if relation_rows:
            _append_jsonl(RELATIONS_JSONL, relation_rows)

        # 追加队列
        if queue_lines:
            with QUEUE.open("a", encoding="utf-8") as qf:
                for ql in queue_lines:
                    qf.write(ql)

        # 评估低置信度
        if args.confidence in cfg.get("quality_rules", {}).get("low_confidence_values", []):
            with QUEUE.open("a", encoding="utf-8") as qf:
                qf.write(f"\n| {date} | `{existing['id']}` | 低置信度 | 该条目标记为低置信度 | 补充来源与证据后复核 | 待处理 |")

        # 更新索引中对应条目
        for ic in index_cards:
            if ic.get("id") == existing["id"]:
                ic["type"] = args.type
                ic["source"] = args.source
                ic["keywords"] = [k.strip() for k in args.keywords.split(",") if k.strip()] if args.keywords else []
                break
        save_index(index_cards)

        print(json.dumps({
            "ok": True,
            "action": action,
            "id": existing["id"],
            "related_count": len(relation_rows),
            "tag_max": max_tags,
        }, ensure_ascii=False))
        return

    # 新建卡片
    if use_notion_dedupe:
        body_md = wrap_notion_export(content)
    else:
        body_md = content

    card = {
        "id": new_id,
        "title": args.title,
        "type": args.type,
        "confidence": args.confidence,
        "tags": ",".join(inferred_tags),
        "keywords": args.keywords,
        "source": args.source,
        "source_url": args.source_url,
        "source_mode": args.source_mode,
        "source_path": args.source_path,
        "notion_page_id": notion_display if notion_display else "",
        "lifecycle": "阶段",
        "created": date,
        "body_md": body_md,
    }

    cards_list.append(card)
    _write_jsonl(CARDS_JSONL, cards_list)

    # 更新 library_index.json
    index_cards.append({
        "id": new_id,
        "title": args.title,
        "type": args.type,
        "source": args.source,
        "date": date,
        "keywords": [k.strip() for k in args.keywords.split(",") if k.strip()] if args.keywords else [],
        "confidence": args.confidence,
    })
    save_index(index_cards)

    # 计算关联
    relation_rows, queue_lines = compute_relations(
        new_id=new_id,
        new_body=body_md,
        new_title=args.title,
        new_tags=inferred_tags,
        existing_cards=[c for c in cards_list if c.get("id") != new_id],
        date=date,
    )

    if relation_rows:
        _append_jsonl(RELATIONS_JSONL, relation_rows)

    if queue_lines:
        with QUEUE.open("a", encoding="utf-8") as qf:
            for ql in queue_lines:
                qf.write(ql)

    if args.confidence in cfg.get("quality_rules", {}).get("low_confidence_values", []):
        with QUEUE.open("a", encoding="utf-8") as qf:
            qf.write(f"\n| {date} | `{new_id}` | 低置信度 | 该条目标记为低置信度 | 补充来源与证据后复核 | 待处理 |")

    print(json.dumps({
        "ok": True,
        "action": action,
        "id": new_id,
        "related_count": len(relation_rows),
        "tag_max": max_tags,
    }, ensure_ascii=False))


def _write_jsonl(path: Path, rows: list[dict]) -> None:
    """覆盖写入 JSONL。"""
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8") as f:
        for row in rows:
            f.write(json.dumps(row, ensure_ascii=False) + "\n")


def _append_jsonl(path: Path, rows: list[dict]) -> None:
    """追加写入 JSONL。"""
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("a", encoding="utf-8") as f:
        for row in rows:
            f.write(json.dumps(row, ensure_ascii=False) + "\n")


if __name__ == "__main__":
    main()
